src/lib/storage/chunk_encoder.cpp

```diff
@@ -12,42 +12,118 @@
 #include "statistics/chunk_statistics/segment_statistics.hpp"
 #include "storage/base_encoded_segment.hpp"
 #include "storage/segment_encoding_utils.hpp"
+#include "storage/segment_iterables/any_segment_iterable.hpp"
+#include "storage/value_segment.hpp"
 #include "utils/assert.hpp"
 
 namespace opossum {
 
+/**
+ * Function takes an arbitrary segment and (re-)encodes it. This reencoding can both mean that
+ * an encoded segment is being recreated as an uncompressed segment (created on the fly via
+ * an AnySegmentIterable) or that another encoding is applied (via the segment encoding utils).
+ */
+std::shared_ptr<BaseSegment> ChunkEncoder::encode_segment(const std::shared_ptr<BaseSegment>& segment,
+                                                          const DataType data_type,
+                                                          const SegmentEncodingSpec& encoding_spec) {
+  std::shared_ptr<BaseSegment> result;
+  resolve_data_type(data_type, [&](auto type) {
+    using ColumnDataType = typename decltype(type)::type;
+
+    // TODO(anyone): After #1489, build segment statistics in encode_segment() instead of encode_chunk()
+    // and store them within the segment instead of a chunk-owned list of statistics.
+
+    if (const auto reference_segment = std::dynamic_pointer_cast<const ReferenceSegment>(segment)) {
+      Fail("Reference segments cannot be encoded.");
+    }
+
+    // Check if early exit is possible when passed segment is currently unencoded and the same is requested.
+    const auto unencoded_segment = std::dynamic_pointer_cast<const ValueSegment<ColumnDataType>>(segment);
+    if (unencoded_segment && encoding_spec.encoding_type == EncodingType::Unencoded) {
+      result = segment;
+      return;
+    }
+
+    // Check for early exit, when requested segment encoding is already being used for the passed segment.
+    const auto encoded_segment = std::dynamic_pointer_cast<const BaseEncodedSegment>(segment);
+    if (encoded_segment && encoded_segment->encoding_type() == encoding_spec.encoding_type) {
+      // Encoded segments do not need to be reencoded when the requested encoding is already present and either
+      // (i) no vector compression is given or (ii) the vector encoding is also already present. Hence, with
+      // {EncodingType::Dictionary} being in place, {EncodingType::Dictionary, VectorCompressionType::SimdBp128}
+      // does not reencode the segment.
+      if (!encoding_spec.vector_compression_type ||
+          (encoding_spec.vector_compression_type && encoded_segment->compressed_vector_type() &&
+           *encoding_spec.vector_compression_type ==
+               parent_vector_compression_type(*encoded_segment->compressed_vector_type()))) {
+        result = segment;
+        return;
+      }
+    }
+
+    // In case of unencoded re-encoding, an any segment iterable is used to manually setup
```

Comment:
Use the actual class name (`AnySegmentIterable`) so that it is easier to find should we ever want to rename it